To support codes via reference i.e. MedicationAdministration which

geloro94 opened the pull request:
To support codes via reference i.e. MedicationAdministration which
references a Medication, and we are interested in the code of the medication
ReferenceCriterion is introduced. The necessary Define statements to refer to
are collected using the Container class and are displayed in the Undefined
Context above the Patient Context. This is a first working Version, with
hardcoded path for MedicationAdministration. Long term an additional mapping
type is required, or expressing support for more complex valueFhirPath like:
MS.medication.as(Reference).resolve().as(Medication).code.
Make sure you have a mapping version with valueFhirPath=Medication.reference
for MedicationAdministration.

geloro94 committed:
To support codes via reference i.e. MedicationAdministration which
references a Medication, and we are interested in the code of the medication
ReferenceCriterion is introduced. The necessary Define statements to refer to
are collected using the Container class and are displayed in the Undefined
Context above the Patient Context. This is a first working Version, with
hardcoded path for MedicationAdministration. Long term an additional mapping
type is required, or expressing support for more complex valueFhirPath like:
MS.medication.as(Reference).resolve().as(Medication).code.
Make sure you have a mapping version with valueFhirPath=Medication.reference
for MedicationAdministration.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/Context.java:
```diff
@@ -0,0 +1,30 @@
+
+
+  public Context {
+    expressionDefinitions = List.copyOf(expressionDefinitions);
```
Please add a `requireNonNull` for `contextName`.

alexanderkiel replied:
Please add a `requireNonNull` for `contextName` here.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/Context.java:
```diff
@@ -0,0 +1,30 @@
+import de.numcodex.sq2cql.PrintContext;
+import java.util.List;
+
+public record Context(String contextName, List<ExpressionDefinition> expressionDefinitions) {
```
IMHO `name` is sufficient, because the class is already called `Context`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/QueryExpression.java:
```diff
@@ -1,25 +1,39 @@
 import static java.util.Objects.requireNonNull;
 
-public record QueryExpression(SourceClause sourceClause, WhereClause whereClause) implements Expression {
+public record QueryExpression(SourceClause sourceClause, WhereClause whereClause, ReturnClause returnClause) implements Expression {
```
If either the WhereClause or the ReturnClause can be non-null here, one single Clause field might be a better idea. Then also print could say the same as before the PR.

geloro94 replied:
For our current usage either of the clauses has to be null. In the future, this won't be the case as a query can combine all the clauses therefore, this version should already address future uses without significant drawbacks.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/RetrieveExpression.java:
```diff
@@ -17,6 +21,9 @@ public static RetrieveExpression of(String resourceType, Expression terminology)
 
     @Override
     public String print(PrintContext printContext) {
+        if(Objects.isNull(terminology)) {
```
A simple `terminology == null` is better here. `Objects.isNull` exists for the use as predicate. Also I would use the ternary operator with a single return here.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/ReturnClause.java:
```diff
@@ -0,0 +1,20 @@
+    return new ReturnClause(expression);
+  }
+
+  public String toCql(PrintContext printContext) {
```
`toCql` should be called `print`.  Sorry that is a leftover in all clauses. It's right that clauses aren't expressions but `toCql` is for SQ criteria and print is for CQL artefacts. Please also change the `WhereClause` and `SourceClause`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/AbstractCriterion.java:
```diff
@@ -28,133 +29,174 @@
+   */
+  static IdentifierExpression referenceName(MappingContext mappingContext,
+      TermCode termCode) {
+    return IdentifierExpression.of("\"%s\"".formatted(termCode.display()));
```
The IdentifierExpression should do the quoting on `print` depending wether any unsupported chars are in the string.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/ReferenceCriterion.java:
```diff
@@ -0,0 +1,52 @@
+
+  }
+
+  public static ReferenceCriterion from(AbstractCriterion criterion, TermCode referencedTermCode) {
```
This method should be called `of`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/SourceClause.java:
```diff
@@ -4,19 +4,23 @@
     public String toCql(PrintContext printContext) {
         assert printContext.precedence() == 0;
-        return "from %s %s".formatted(querySource.print(printContext), alias.print(printContext));
+        return "%s%s %s".formatted(from ? "from " : "", querySource.print(printContext.increase()), alias.print(printContext));
```
I see no need to remove the `from` here.

geloro94 replied:
Correct, and implementations even might require it:
https://cql.hl7.org/03-developersguide.html#query-syntax-options

geloro94 replied:
Correct, and implementations even might require it:
https://cql.hl7.org/03-developersguide.html#query-syntax-options

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/ReferenceCriterion.java:
```diff
@@ -0,0 +1,52 @@
+    var retrieveExpr = RetrieveExpression.of("Medication", codeSelector.get());
+    var query = QueryExpression.of(SourceClause.of(retrieveExpr, retrieveExpr.alias()),
+        ReturnClause.of(IdentifierExpression.of("'Medication/' + M.id")));
+    var defineExpr = new ExpressionDefinition("\"%s\"".formatted(referencedTermCode.display()), query);
```
Also, please perform the quoting on print. I also Would add a "Ref" to the name because we return the reference from that expression and not the resource itself.

alexanderkiel replied:
Also this isn't an expression, so `defineExpr` is not an appropriate name.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Container.java:
```diff
@@ -58,7 +61,21 @@ public static <T> Container<T> empty() {
+     *
+     * @param expression            the expression being hold
+     * @param codeSystemDefinitions the code system definition being hold
+     * @param <T>                   the type of the expression
```
Please move this line one down in order to have all arguments together.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Container.java:
```diff
@@ -58,7 +61,21 @@ public static <T> Container<T> empty() {
+     * @throws NullPointerException if {@code expression} is null
+     */
+    public static <T> Container<T> of(T expression, Set<CodeSystemDefinition> codeSystemDefinitions, Set<ExpressionDefinition> referencedDefinitions) {
+        return new Container<>(requireNonNull(expression), codeSystemDefinitions, referencedDefinitions);
```
Please copy the sets here.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Container.java:
```diff
@@ -98,6 +115,13 @@ public Set<CodeSystemDefinition> getCodeSystemDefinitions() {
     }
 
+    /**
+     * Returns the referenced Codes
```
> Returns the definitions that are referenced within the patient context.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/AbstractCriterion.java:
```diff
@@ -28,133 +29,174 @@
+  /**
+   * Returns the name of a coding as stringliteralExpression to use as reference
+   */
+  static IdentifierExpression referenceName(MappingContext mappingContext,
```
Please remove the unused MappingContext.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/ReferenceCriterion.java:
```diff
@@ -0,0 +1,52 @@
+    var codeSelector = codeSelector(mappingContext, referencedTermCode).getExpression();
+    var retrieveExpr = RetrieveExpression.of("Medication", codeSelector.get());
+    var query = QueryExpression.of(SourceClause.of(retrieveExpr, retrieveExpr.alias()),
+        ReturnClause.of(IdentifierExpression.of("'Medication/' + M.id")));
```
That is no IdentifierExpression. Please Introduce a RawExpression if it's to much work to build the Expression properly.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/AbstractCriterion.java:
```diff
@@ -28,133 +29,174 @@
+    if ("Patient".equals(mapping.resourceType())) {
+      return valueAndModifierExpr(mappingContext, mapping, PATIENT);
+    } else if ("MedicationAdministration".equals(mapping.resourceType())) {
+      return retrieveResourceTypeExpr(mappingContext, termCode).flatMap(retrieveExpr -> {
```
This is just a `MedicationAdministration` retrieve expression. There is no need for this extra `retrieveResourceTypeExpr` method.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/AbstractCriterion.java:
```diff
@@ -28,133 +29,174 @@
+      return valueAndModifierExpr(mappingContext, mapping, PATIENT);
+    } else if ("MedicationAdministration".equals(mapping.resourceType())) {
+      return retrieveResourceTypeExpr(mappingContext, termCode).flatMap(retrieveExpr -> {
+        var self = ReferenceCriterion.from(this, termCode);
```
I see that you use some code from AbstractCriterion here. But at the end, there is no ReferenceCriterion. Please put the code generating the expression container into a class which isn't a criterion.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/structured_query/ReferenceCriterion.java:
```diff
@@ -0,0 +1,52 @@
+  Container<BooleanExpression> valueExpr(MappingContext mappingContext, Mapping mapping,
+      IdentifierExpression identifier) {
+    var valueExpr = InvocationExpression.of(identifier, mapping.valueFhirPath());
+    var codeSelector = codeSelector(mappingContext, referencedTermCode).getExpression();
```
By throwing the container away, you loose the code system definition, the code selector needs. This only works in your example, because you generate the code selector on MedicationAdministration in `retrieveResourceTypeExpr`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Translator.java:
```diff
@@ -13,111 +14,133 @@
+
+  private static Library inclusionOnlyLibrary(Container<BooleanExpression> inclusionExpr) {
+    assert inclusionExpr.getExpression().isPresent();
+    var UndefinedContext = Context.of("Unfiltered",
```
Please start vars in lowercase.

alexanderkiel replied:
And it's `unfiltered` and not `undefined`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Translator.java:
```diff
@@ -13,111 +14,133 @@
+        List.copyOf(inclusionExpr.getReferencedDefinitions()));
+    var PatientContext = Context.of("Patient", List.of(
+        ExpressionDefinition.of("InInitialPopulation", inclusionExpr.getExpression().get())));
+    if (UndefinedContext.expressionDefinitions().isEmpty()) {
```
Just filter out empty contexts on printing or don't allow empty contexts at all.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/Context.java:
```diff
@@ -0,0 +1,30 @@
+    expressionDefinitions = List.copyOf(expressionDefinitions);
+  }
+
+  public static Context of() {
```
Please remove this unused method.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/Context.java:
```diff
@@ -0,0 +1,30 @@
+    return new Context(contextName, expressionDefinitions);
+  }
+
+  public String print(PrintContext printContext) {
```
Please remove the print context here and at the library level, because it's always the ZERO print context. We only need the print context at expression level.

geloro94 committed:
Refactoring addressing:
How reference definitions are added to Container
Naming
Adding Expressions to create return clause 'Medication/' + M.id
Removed PrintContext from Library and Context

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/AdditionExpressionTerm.java:
```diff
@@ -0,0 +1,36 @@
+    ExpressionTerm {
+
+  public AdditionExpressionTerm {
+    List.copyOf(expressions);
```
Has to be `expressions = List.copyOf(expressions)`

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/AdditionExpressionTerm.java:
```diff
@@ -0,0 +1,36 @@
+    List.copyOf(expressions);
+  }
+
+  public static final int PRECEDENCE = 16;
```
Please move that constant above the constructor. We have constants usually as first items in a class.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/IdentifierExpression.java:
```diff
@@ -6,16 +6,17 @@
-    }
+  @Override
+  public String print(PrintContext printContext) {
+    return identifier.matches("([A-Za-z]|'_')([A-Za-z0-9]|'_')*") ? identifier
```
Please add an explicit test in a new class IdentifierExpressionTest.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/model/cql/QueryExpression.java:
```diff
@@ -1,25 +1,39 @@
-        var wherePrintContext = sourcePrintContext.increase();
-        return printContext.parenthesizeZero("%s\n%s%s".formatted(sourceClause.toCql(sourcePrintContext),
-                wherePrintContext.getIndent(), whereClause.toCql(wherePrintContext)));
+        if (!Objects.isNull(whereClause)) {
```
Please use just `whereClause == null`.

alexanderkiel commented on src/main/java/de/numcodex/sq2cql/Container.java:
```diff
@@ -13,115 +15,179 @@
+   * @return a container
+   * @throws NullPointerException if {@code expression} is null
+   */
+  public static <T extends Expression> Container<T> of(T expression,
```
Can be private.

alexanderkiel reviewed (approved):
Comment for GitHub

geloro94 committed:
Added IdentifierExpressionTest
